File: ToolGood.Words/internals/BaseSearch.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ToolGood.Words.internals
{
    public abstract class BaseSearch
    {
        //protected TrieNode _root = new TrieNode();
        protected TrieNode[] _first = new TrieNode[char.MaxValue + 1];

        /// <summary>
        /// 设置关键字
        /// </summary>
        /// <param name="_keywords">关键字列表</param>
        public virtual void SetKeywords(ICollection<string> _keywords)
        {
            var first = new TrieNode[char.MaxValue + 1];
            var root = new TrieNode();

            foreach (var p in _keywords) {
                if (string.IsNullOrEmpty(p)) continue;

                var nd = _first[p[0]];
                if (nd == null) {
                    nd = root.Add(p[0]);
                    first[p[0]] = nd;
                }
                for (int i = 1; i < p.Length; i++) {
                    nd = nd.Add(p[i]);
                }
                nd.SetResults(p);
            }
            this._first = first;// root.ToArray();

            Dictionary<TrieNode, TrieNode> links = new Dictionary<TrieNode, TrieNode>();
            foreach (var item in root.m_values) {
                TryLinks(item.Value, null, links);
            }

            foreach (var item in links) {
                item.Key.Merge(item.Value, links);
            }

            //_root = root;
        }

        private void TryLinks(TrieNode node, TrieNode node2, Dictionary<TrieNode, TrieNode> links)
        {
            foreach (var item in node.m_values) {
                TrieNode tn = null;
                if (node2 == null) {
                    tn = _first[item.Key];
                    if (tn != null) {
                        links[item.Value] = tn;
                    }
                } else if (node2.TryGetValue(item.Key, out tn)) {
                    links[item.Value] = tn;
                }
                TryLinks(item.Value, tn, links);
            }
        }


    }
}

File: java/ToolGood/Words/internals/BaseSearch.java
package ToolGood.Words.internals;

import java.util.Hashtable;
import java.util.List;
import java.util.Map;

import org.springframework.util.StringUtils;

public abstract class BaseSearch {
    protected TrieNode[] _first = new TrieNode[Character.MAX_VALUE + 1];


    public void SetKeywords(List<String> _keywords)
    {
        TrieNode[] first =  new TrieNode[Character.MAX_VALUE + 1];
        TrieNode root = new TrieNode();

        _keywords.forEach(p->{
            if(StringUtils.isEmpty(p)==false){ 
                TrieNode nd = _first[ p.charAt(0)];
                if (nd == null) {
                    nd = root.Add(p.charAt(0));
                    first[p.charAt(0)] = nd;
                }
                for (int i = 1; i < p.length(); i++) {
                    nd = nd.Add(p.charAt(i));
                }
                nd.SetResults(p);
            }
        });
 
        this._first = first;// root.ToArray();

        Map<TrieNode, TrieNode> links = new Hashtable<TrieNode, TrieNode>();
        root.m_values.forEach((k,v)->{
            TryLinks(v, null, links);
        });
        links.forEach((k,v)->{
            k.Merge(v, links);
        });
    }

    private void TryLinks(TrieNode node, TrieNode node2, Map<TrieNode, TrieNode> links)
    {
        node.m_values.forEach((Key,Value)->{
            TrieNode tn = null;
            if (node2 == null) {
                tn = _first[Key];
                if (tn != null) {
                    links.put(Value, tn);
                }
            } else if (node2.TryGetValue(Key, tn)) {
                links.put(Value, tn);
            }
            TryLinks(Value, tn, links);
        });
    }

}